### PR TITLE
Add unit tests for TimeSeriesProperty::cloneWithTimeShift

### PR DESCRIPTION
**Pull request by TTitcombe**
**Description of work.**
While investigating a bug caused by adding event workspaces containing erroneous pulse timestamps, it was necessary to test `TimeSeriesProperty::cloneWithTimeShift`.

As it turns out, there were no existing unit tests for this method. Now there are.

I've also included a note on the bug I was investigating, to avoid the next person wasting time looking there. (I originally thought it could have somehow been caused by shifting a time over unix time 0, however this is not the case)

**To test:**
Code review and check that `TimeSeriesPropertyTest` passes

*This does not require release notes* because **internal change to test code**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Kernel/test/TimeSeriesPropertyTest.h [hunk 1]
```diff
@@ -1345,6 +1345,47 @@ class TimeSeriesPropertyTest : public CxxTest::TestSuite {
     return;
   }
 
+  /*
+   * Test cloneWithTimeShift
+   */
+  void test_cloneWithTimeShift() {
+    TimeSeriesProperty<int> *time_series =
```
Comment (DTasev) [inline Framework/Kernel/test/TimeSeriesPropertyTest.h]: Can these be changed to `auto`?

File: Framework/Kernel/test/TimeSeriesPropertyTest.h [hunk 2]
```diff
@@ -1345,6 +1345,47 @@ class TimeSeriesPropertyTest : public CxxTest::TestSuite {
...
+        new TimeSeriesProperty<int>("IntUnixTest");
+    time_series->addValue("2019-02-10T16:17:00", 1);
+
+    TimeSeriesProperty<int> *time_series_small_shift =
```
Comment (DTasev) [inline Framework/Kernel/test/TimeSeriesPropertyTest.h]: These _could_ be changed to unique_ptr, but the init is kind of ugly:
`std::unique_ptr<TimeSeriesProperty<int>>(<TimeSeriesProperty<int> *>(time_series->cloneWithTimeShift(100.)));`.

Up to you if you want to change them or leave them. I don't mind either way

File: Framework/Kernel/test/TimeSeriesPropertyTest.h [hunk 3]
```diff
@@ -1345,6 +1345,47 @@ class TimeSeriesPropertyTest : public CxxTest::TestSuite {
...
+        dynamic_cast<TimeSeriesProperty<int> *>(
+            time_series->cloneWithTimeShift(100.));
+    std::vector<DateAndTime> small_shift_times =
```
Comment (DTasev) [inline Framework/Kernel/test/TimeSeriesPropertyTest.h]: I think these can be captured by `const auto &`, as I don't think they are be deleted before the assertions

**Review by dtasev — CHANGES_REQUESTED** [review]
Good addition! Tests pass locally, just a few minor comments

File: Framework/Kernel/test/TimeSeriesPropertyTest.h [hunk 4]
```diff
@@ -1345,6 +1345,47 @@ class TimeSeriesPropertyTest : public CxxTest::TestSuite {
     return;
   }
 
+  /*
+   * Test cloneWithTimeShift
+   */
+  void test_cloneWithTimeShift() {
+    TimeSeriesProperty<int> *time_series =
+        new TimeSeriesProperty<int>("IntUnixTest");
```
[the remark this answers is not in the document]
Comment (TTitcombe) [reply Framework/Kernel/test/TimeSeriesPropertyTest.h]: Regarding all the comments about using modern c++, I've tried to keep the new tests in the local style. This file was quite old so will probably need to be updated at some point.
Are we working under the basis of using modern c++ where possible, and update the rest of the file at a later date?

[the remark this answers is not in the document]
Comment (DTasev) [reply Framework/Kernel/test/TimeSeriesPropertyTest.h]: Personally, I do as you said: use modern C++ where possible, and update the rest later. 

This test class is pretty big so it shouldn't be part of this PR, but maybe a maintenance task

[the remark this answers is not in the document]
Comment (TTitcombe) [reply Framework/Kernel/test/TimeSeriesPropertyTest.h]: I think we already have a maintenance task regarding this, but I'll open one if not

File: Framework/Kernel/test/TimeSeriesPropertyTest.h (hunk 2 above)
Comment (TTitcombe) [reply Framework/Kernel/test/TimeSeriesPropertyTest.h]: I've decided to leave them as it's slightly cleaner

**Review by dtasev — APPROVED** [review]

**Comment by martyngigg** [comment]
RHEL7 agent went offline. Unit tests passed.

**Pull request merged**